content/en/blog/_posts/2024-09-25-swap-ga-graduation.md

```diff
@@ -0,0 +1,369 @@
+the total physical swap capacity of the machine.
+
+### NFD (Node Feature Discovery)
+
+[Node Feature Discovery](https://github.com/kubernetes-sigs/node-feature-discovery)
+is a Kubernetes addon for detecting hardware features and configuration.
+It can be utilized to discover which nodes are provisioned with swap.
+
+As an example, to figure out if a certain node is provisioned with swap,
+the following command can be executed while replacing `<NODE-NAME>` with the node's name:
+```shell
+> k get node <NODE-NAME> -o yaml | grep "feature.node.kubernetes.io/memory-swap"
+      feature.node.kubernetes.io/memory-swap: "true"
+```
+
+## Caveats
+
+Having swap available on a system reduces predictability.
+While swap can enhance performance by making more RAM available, swapping data
+back to memory is a heavy operation, sometimes slower by many orders of magnitude,
+which can cause unexpected performance regressions.
+Furthermore, swap changes a system's behaviour under memory pressure.
+Since enabling swap permits greater memory usage for workloads in Kubernetes that
+cannot be predictably accounted for, it also increases the risk of noisy neighbours
+and unexpected packing configurations, as the scheduler currently does not account for swap memory usage.
+
+The performance of a node with swap memory enabled depends on the underlying physical storage.
+When swap memory is in use, performance will be significantly worse in an I/O
+operations per second (IOPS) constrained environment, such as a cloud VM with
+I/O throttling, when compared to faster storage mediums like solid-state drives
+or NVMe.
+As swap might cause IO pressure, it is recommended to give a higher IO latency
+priority to system critical daemons. See the relevant section in the
+[best-practices](#best-practices) section below.
+
+### Memory-backed volumes
+
+On Linux nodes, memory-backed volumes (like [Secrets](/docs/concepts/configuration/secret/)
+and [emptyDirs](/docs/concepts/storage/volumes/#emptydir) with `medium: Memory`)
+are implemented with as `tmpfs` mount.
+The contents of such volumes should remain in memory at all times, hence should
+not be swapped to disk.
+To ensure the contents of such volumes remain in memory, the `noswap` tmpfs option
+is being used.
+
+The Linux kernel officially supports the `noswap` option from version 6.3 (more info
+can be found [here](/docs/reference/node/kernel-version-requirements/#requirements-other)).
+However, the different distributions often choose to backport this option to older
+Linux versions as well.
+
+In order to understand if the node supports the `noswap` option, Kubelet will do the following:
+* If the kernel's version is above 6.3 then the `noswap` option will be assumed to be supported.
+* Otherwise, kubelet would try to mount a dummy tmpfs with the `noswap` option at startup.
+  If kubelet fails with an error indicating of an unknown option, `noswap` will be assumed
+  to not be supported, hence will not be used.
```

Comment:
What happens once the kubelet detects the lack of `noswap` support? Does the kubelet refuse to start? Do you get an Event against the node?

Reply:
I'd kind of hope to see a kubelet configuration option such as `allowInsecureMemoryVolumes` that you have to turn on if using an older kernel. Is there anything like that?

Reply:
> What happens once the kubelet detects the lack of `noswap` support? Does the kubelet refuse to start? Do you get an Event against the node?

When `noswap` is not supported, kubelet would start with a log entry warning.

There were long discussions regarding this topic (for example, look [here](https://github.com/kubernetes/kubernetes/issues/105978#issuecomment-2078246497) for more information). In short, this behavior is reasonable since most distribution backported this mount option to much older versions. The vast majority of systems would have this supported, and there is documentation and warnings for the systems that don't, that anyhow are likely to update their Kernel version in a reasonable timeframe.

Reply:
We might want to call out the risk of enabling swap on an old kernel; even if it's unlikely, some people may face important opportunities to get things really wrong. Helping them they that they haven't and emphasizing how important it is to check? I recommend it.

Reply:
> When `noswap` is not supported, kubelet would start with a log entry warning.

OK, but: don't tell the reviewers. Tell the readers.

Reply:
> Helping them they that they haven't and emphasizing how important it is to check?

So you suggest adding a node-level event in addition to the kubelet log? I think that might be reasonable.

Reply:
> > When `noswap` is not supported, kubelet would start with a log entry warning.
> 
> OK, but: don't tell the reviewers. Tell the readers.

Added.

Reply:
The Node-level event would be a nice improvement, but whether it's in or not without it let's clearly call out how to check what's going on.